### ISSUE TITLE
Lib: escape_url() does not correctly escape UTF8 strings

### ISSUE THREAD

**Issue by ChristianBeer** [BOINC/boinc#1951]
The [escape_url()](https://github.com/BOINC/boinc/blob/e712c7af9ac5b4f9c16d691e656788a2112cefe3/lib/url.cpp#L161) function does not correctly escape UTF8 characters (umlauts) which in case of the [create_account RPC](https://github.com/BOINC/boinc/blob/e712c7af9ac5b4f9c16d691e656788a2112cefe3/client/acct_setup.cpp#L154) leads to an invalid username error and no account is created.

```
before: aeiou äëïöü áéíóú àèìòù
after: aeiou%20%-#%+$%-#%++%-#%+/%-#%,&%-#%,,%20%-#%+!%-#%+)%-#%+-%-#%,#%-#%,*%20%-#%*0%-#%+(%-#%+,%-#%,"%-#%,)
correct: aeiou%20%C3%A4%C3%AB%C3%AF%C3%B6%C3%BC%20%C3%A1%C3%A9%C3%AD%C3%B3%C3%BA%20%C3%A0%C3%A8%C3%AC%C3%B2%C3%B9
```
Can be tested here: https://www.w3schools.com/tags/ref_urlencode.asp

I see two solutions:

1. fix the existing function to properly recognize UTF8 which in my experience only makes the function more complex and we still might not catch every case
1. find a library that does this kind of stuff and has proven itself to cover all the cases

I also think that this should be part of the next 7.6 maintenance release.

**Comment by ChristianBeer** [comment]
It turns out it can be as simple as:
```
#include <curl/curl.h>

void escape_url(const char *in, char*out, int out_size) {
    CURL *curl = curl_easy_init();
    if(curl) {
        char *output = curl_easy_escape(curl, in, 0);
        if(output) {
            strcpy(out, output);
            curl_free(output);
        }
    }
    curl_easy_cleanup(curl);
}
```
Some buffer overflow checks are still needed. And this would require curl as a dependency for the library but it would solve the problem.

**Comment by JuhaSointusalo** [comment]
I did a bit of testing. The problem isn't UTF8 as such but characters (or code units) with value >127.

In C `char` type can be signed or unsigned and is left for implementation to decide. GCC and Visual C++ on x86 have signed `char`, GCC on ARM has unsigned `char`. The algorithm in `c2x` appears to depend on `unsigned char` and produces garbage with `signed char`.

It looks like changing [this line](https://github.com/BOINC/boinc/blob/e712c7af9ac5b4f9c16d691e656788a2112cefe3/lib/url.cpp#L120) to read

    unsigned char num = atoi(what);

is enough to fix it but maybe it would be better to change everything to `unsigned char` and in any case it really needs proper testing. `x2c` (and `unescape_url`) appears to be unaffected by the actual type but should be tested properly as well.

Also, `escape_url` uses `isalnum` to decide if it needs to encode the character. That works when using `C` locale but not so much in many other locales. It might be better to change it to explicitly test for `[a-zA-Z0-9]`.

**Comment by CharlieFenton** [comment]
It is also easy to temporarily set the locale to C and then back to what it was. This is already done in many place in the code. See the definition of SET_LOCALE in lib/gui_rpc_client.h and its use in lib/gui_rpc_client_ops.cpp and clientgui/DlgDianosticLogFlags.cpp.

(You can ignore the part about OSX versions < 10.4 because we now only support OS 10.6 and later.)

**Comment by CharlieFenton** [comment]
Will explicitly testing for [a-zA-Z0-9] work for Unicode alpha characters with accent marks? 

Would it be better to use a locale-specific function such as isalnum_l(int c, locale_t loc) or iswalnum_l(wint_t wc, locale_t loc)? I know these are available on Macintosh; I don't know how widely they are available on other platforms.

**Comment by ChristianBeer** [comment]
Thanks for also looking into this. Adding the `unsigned` modifier in `c2x()` does indeed do the trick for UTF8 where the highest bit is always set to 1. We could also use a compile flag for GCC to always use unsigned char but that might have other effects we don't want. So it's probably better to just change the instances we need right now.

What the `isalnum()` should really do is filter out all ASCII alphanumeric values (independent of current locale) because they don't need to be escaped. So it might suffice to change locale to "C" in `escape_url()` using this `SET_LOCALE sl` object. There is a `bool isalnum( charT ch, const locale& loc );` function defined in the C++ standard but my understanding is that the library code should mostly be C compatible. Unicode alpha characters with accent marks will always be represented as 2 char where both are >127 to distinguish them from ASCII. So calling `isalnum()` with an `unsigned char` (1 byte) will always get us the correct result. Only if you call it with an `int` (or anything >=2 byte) this becomes a problem if the value is a valid UTF8 encoding of an alphanumeric character and the current locale is not "C". In this case the URL escaping would fail.

I'm preparing a pull request now.

**Comment by ChristianBeer** [comment]
I copied the relevant functions into a separate file and used this main loop for testing:

```
#include <stdio.h>
#include <string>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>

using std::string;
#define strlcpy strncpy

// functions go here

int main(int argc, char**argv) {
    string temp1 = "aeiou äëïöü áéíóú àèìòù";
    string temp2 = "Epäjärjestelmällistyttämättömyydellänsäkäänköhän";

    printf("before: %s\n", temp1.c_str());
    escape_url(temp1);
    printf("after : %s\n", temp1.c_str());
    unescape_url(temp1);
    printf("back  : %s\n", temp1.c_str());
    printf("before: %s\n", temp2.c_str());
    escape_url(temp2);
    printf("after : %s\n", temp2.c_str());
    unescape_url(temp2);
    printf("back  : %s\n", temp2.c_str());

    return 0;
}
```

**Comment by davidpanderson** [comment]
I checked in a change (basically Juha's) that passes Christian's test, at least in the C locale.

**Comment by RichardHaselgrove** [comment]
That must be ff7633a65def104b81a82b839f98fbe4b383f42f - on test list for v7.10

**Comment by AenBleidd** [comment]
@RichardHaselgrove, was this tested during 7.10 release test phase? Could this issue be closed?

**Comment by Uplinger** [comment]
Closing because this has a unit test on it and appears to be fixed.
